xrootd/client.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package xrootd // import "go-hep.org/x/hep/xrootd"
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+
+	"go-hep.org/x/hep/xrootd/encoder"
+	"go-hep.org/x/hep/xrootd/streammanager"
+)
+
+var logger = log.New(os.Stderr, "xrootd: ", log.LstdFlags)
+
+// A Client to xrootd server
+type Client struct {
+	connection      *net.TCPConn
+	sm              *streammanager.StreamManager
+	protocolVersion int32
+}
+
+type serverError struct {
+	Code    int32
+	Message string
+}
+
+func (err serverError) Error() string {
+	return fmt.Sprintf("Server error %d: %s", err.Code, err.Message)
+}
+
+const responseHeaderSize = 2 + 2 + 4
+
+type responseHeader struct {
+	StreamID   streammanager.StreamID
+	Status     uint16
+	DataLength int32
+}
+
+// New creates a client to xrootd server at address
+func New(ctx context.Context, address string) (*Client, error) {
+	conn, err := createTCPConnection(address)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{conn, streammanager.New(), 0}
+
+	go client.consume()
+
+	err = client.handshake(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func createTCPConnection(address string) (connection *net.TCPConn, err error) {
+	tcpAddr, err := net.ResolveTCPAddr("tcp4", address)
+	if err != nil {
+		return
+	}
+
+	connection, err = net.DialTCP("tcp", nil, tcpAddr)
+	return
+}
+
+func (client *Client) consume() {
+	for {
+		var header = &responseHeader{}
+
+		var headerBytes = make([]byte, responseHeaderSize)
+		if _, err := io.ReadFull(client.connection, headerBytes); err != nil {
+			logger.Panic(err)
```

Comment:
I am not too keen on having the Client panic like so.
could we devise a way to properly propagate this error back to the user ?

(same remark for all the panics below)